text/0000-config-transaction-library.md

```diff
@@ -0,0 +1,84 @@
+---
+layout: default
+title: Create a standalone config transaction library that generates channel config creation and update envelopes
+nav_order: 3
+---
+
+- Feature Name: Create a standalone config transaction library that generates channel config creation and update envelopes
+- Start Date: 2020-03-20
+- RFC PR: (leave this empty)
+- Fabric Component: fabric, fabric-config
+- Fabric Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This feature will implement a consumable standalone library that supports generating configuration envelopes for 
+operations including application and system channel creation, channel configuration updates, and attaching 
+endorsing signatures to generated envelopes.
+
+# Motivation
+[motivation]: #motivation
+
+The motivation for this work is to provide a consistent API that is actually intended for consumption by external
+clients for any operations involving channel config transactions. One example usage case could be a
+CLI tool for creating config updates that takes as flags properties to be updated alongside an existing config transaction.
+This CLI tool could then pass these parameters down to the config package for generating a config update envelope with the computed
+read and write sets. Currently `configtxgen` is available and frequently misused as a tool for these operations, but it was never intended for
+production usage. Much of the existing process for generating configuration updates is both tedious and error prone involving
+decoding a config transaction to JSON with the protolator tool and manually modifying fields prior to encoding back to a protobuf
+and submitting for update.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Work on this consumable library has already begun in [`fabric/pkg/config`](https://github.com/hyperledger/fabric/tree/master/pkg/config) 
+and tracked in [FAB-17628](https://jira.hyperledger.org/browse/FAB-17628). Example usage has been included in the 
+[godocs](https://godoc.org/github.com/hyperledger/fabric/pkg/config#pkg-examples) for this package and can be referenced for external tooling.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+API documentation already exists in the godocs for this package and can be referenced [here](https://godoc.org/github.com/hyperledger/fabric/pkg/config).
```

Comment:
Danny, the below comments are already discussed in the code review. I am just conveying my realization. 

_The client usage flow looks like a state-machine to me. Every API call can result in an `error` and every error result in a `panic`. The user cannot move from one state to another without passing certain criteria. Using these APIs, I am sure that the client may not submit any invalid channel config tx to orderer and it would remove the need for any code that validates the received tx at the orderer node (kidding, still we cannot remove it)._

 I think, for a general-purpose library such as golang `map`, `delete()` not returning error might be okay. After some thought (and some discussions with some maintainers), I realized that this is more of a special purpose library and hence, the `error` returns & state-machine approach are fine. 

We still need to think about whether we should return an `error` or [error type](https://golang.org/doc/effective_go.html#errors) to the library user.

Reply:
Agreed, will have that conversation with the config squad as we wrap up on most of the feature work and move on to cleanup/improvements.